### PR TITLE
[feat][site] PIP-264: Document topic messaging metrics

### PR DESCRIPTION
**Pull request by dragosvictor**
Documentation for https://github.com/apache/pulsar/pull/22467. 

### ✅ Contribution Checklist

- [x] I read the [contribution guide](https://pulsar.apache.org/contribute/document-contribution/)
- [ ] I updated the [versioned docs](https://pulsar.apache.org/contribute/document-contribution/#update-versioned-docs)

### CODE REVIEW

File: docs/reference-metrics-opentelemetry.md [hunk 1]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Type: UpDownCounter
+* Unit: `{message batch}`
+
+#### pulsar.broker.topic.storage.incoming
+The total message batches (entries) read from the storage for this topic.
+* Type: UpDownCounter
+* Unit: `{message batch}`
+
+#### pulsar.broker.topic.compaction.removed.event.count
+The total number of removed events of the compaction.
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:109]: Could we clarify more about this deletion event?

**Comment by heesung-sn** [comment]
It would be great to clarify whether the metrics are accumulative or delta.

File: docs/reference-metrics-opentelemetry.md [hunk 2]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
 
 ## Broker
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:10]: Can we clarify the OTel attributes related to partitioned or non-partitioned topics?

File: docs/reference-metrics-opentelemetry.md [hunk 3]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
 
+### Topic Messaging metrics
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:11]: Do we need to consider Sync vs Async instruments for this `Type`?

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#synchronous-and-asynchronous-instruments

File: docs/reference-metrics-opentelemetry.md [hunk 4]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Unit: `{message batch}`
+
+#### pulsar.broker.topic.storage.incoming
+The total message batches (entries) read from the storage for this topic.
+* Type: UpDownCounter
+* Unit: `{message batch}`
+
+#### pulsar.broker.topic.compaction.removed.event.count
+The total number of removed events of the compaction.
+* Type: UpDownCounter
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:110]: Shouldn't it be Counter, as monotonic?

File: docs/reference-metrics-opentelemetry.md [hunk 5]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Type: Counter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.publish.rate.limit
+The number of times the publish rate limit is triggered.
+* Type: Counter
+* Unit: `{event}`
+
+#### pulsar.broker.topic.consumer.msg.ack
+The total number of message acknowledgments received for this topic.
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:54]: Could we explain when the ack is Cumulative Ack?

File: docs/reference-metrics-opentelemetry.md [hunk 6]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Type: Counter
+* Unit: `{ack}`
+
+#### pulsar.broker.topic.storage.size
+The total storage size of the messages in this topic.
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:59]: Could we clarify if this counts replicas and offloaded data?

File: docs/reference-metrics-opentelemetry.md [hunk 7]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Type: UpDownCounter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.storage.logical.size
+The storage size of topics in the namespace owned by the broker without replicas.
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:64]: Could we clarify if this counts offloaded data?

File: docs/reference-metrics-opentelemetry.md [hunk 8]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total number of messages bytes received for this topic.
+* Type: Counter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.message.outgoing.size
+The total number of messages bytes read from this topic.
+* Type: Counter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.publish.rate.limit
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:48]: Shouldn't it be `pulsar.broker.topic.publish.rate.limit.count`
like
`pulsar.broker.topic.storage.backlog.quota.eviction.count`?

File: docs/reference-metrics-opentelemetry.md [hunk 9]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+#### pulsar.broker.topic.storage.backlog.quota.limit.time
+The time based backlog quota limit for this topic.
+* Type: Gauge
+* Unit: `{second}`
+
+#### pulsar.broker.topic.storage.backlog.quota.eviction.count
+The number of times a backlog was evicted since it has exceeded its quota.
+* Type: Counter
+* Unit: `{eviction}`
+
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:92]: Do we have
`pulsar.broker.topic.storage.backlog.quota.eviction.msg.count`
`pulsar.broker.topic.storage.backlog.quota.eviction.msg.size` metrics?

File: docs/reference-metrics-opentelemetry.md [hunk 10]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+#### pulsar.broker.topic.storage.backlog.age
+The age of the oldest unacknowledged message (backlog).
+* Type: Gauge
+* Unit: `{second}`
+
+#### pulsar.broker.topic.storage.outgoing
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:98]: Perhaps, `pulsar.broker.topic.storage.entry.outgoing.count`?

File: docs/reference-metrics-opentelemetry.md [hunk 11]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total message batches (entries) written to the storage for this topic.
+* Type: UpDownCounter
+* Unit: `{message batch}`
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:101]: Perhaps, entry?

File: docs/reference-metrics-opentelemetry.md [hunk 12]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+
+#### pulsar.broker.topic.storage.incoming
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:103]: Perhaps, `pulsar.broker.topic.storage.entry.incoming.count`?

File: docs/reference-metrics-opentelemetry.md [hunk 13]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total message batches (entries) read from the storage for this topic.
+* Type: UpDownCounter
+* Unit: `{message batch}`
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:106]: Perhaps, entry?

File: docs/reference-metrics-opentelemetry.md [hunk 14]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+
+#### pulsar.broker.topic.compaction.removed.event.count
+The total number of removed events of the compaction.
+* Type: UpDownCounter
+* Unit: `{event}`
+
+#### pulsar.broker.topic.compaction.succeed.count
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:113]: Perhaps, `pulsar.broker.topic.compaction.success.count`

File: docs/reference-metrics-opentelemetry.md [hunk 15]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total number of successes of the compaction.
+* Type: UpDownCounter
+* Unit: `{event}`
+
+#### pulsar.broker.topic.compaction.failed.count
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:118]: Perhaps, pulsar.broker.topic.compaction.failure.count

File: docs/reference-metrics-opentelemetry.md [hunk 16]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total number of failures of the compaction.
+* Type: UpDownCounter
+* Unit: `{event}`
+
+#### pulsar.broker.topic.compaction.duration
+The total time duration of compaction operations on the topic.
+* Type: DoubleUpDownCounter
+* Unit: `{second}`
+
+#### pulsar.broker.topic.compaction.incoming
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:128]: Perhaps, pulsar.broker.topic.compaction.incoming.size?

File: docs/reference-metrics-opentelemetry.md [hunk 17]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total count of bytes read by the compaction process for this topic.
+* Type: UpDownCounter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.compaction.outgoing
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:133]: perhaps, pulsar.broker.topic.compaction.outgoing.size?

File: docs/reference-metrics-opentelemetry.md [hunk 18]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total count of bytes written by the compaction process for this topic.
+* Type: UpDownCounter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.compaction.compacted.entries.count
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:138]: perhaps `pulsar.broker.topic.compaction.compacted.entry.count`

File: docs/reference-metrics-opentelemetry.md [hunk 19]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total number of compacted entries.
+* Type: UpDownCounter
+* Unit: `{entry}`
+
+#### pulsar.broker.topic.compaction.compacted.entries.size
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:143]: perhaps `pulsar.broker.topic.compaction.compacted.entry.size`

File: docs/reference-metrics-opentelemetry.md [hunk 20]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total size of the compacted entries.
+* Type: UpDownCounter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.transaction
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:148]: perhaps, pulsar.broker.topic.transaction.count

File: docs/reference-metrics-opentelemetry.md [hunk 21]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The number of transactions on this topic.
+* Type: UpDownCounter
+* Unit: `{transaction}`
+
+#### pulsar.broker.topic.subscription.delayed
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:153]: perhaps, pulsar.broker.topic.subscription.delayed.entry.count

File: docs/reference-metrics-opentelemetry.md [hunk 22]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
 
 ## Broker
 
+### Topic Messaging metrics
+
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:12]: What's the metric naming convention of OTel? Should we put `count`/`size` suffix?

File: docs/reference-metrics-opentelemetry.md [hunk 23]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+The total number of messages bytes read from this topic.
+* Type: Counter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.publish.rate.limit
+The number of times the publish rate limit is triggered.
+* Type: Counter
+* Unit: `{event}`
+
+#### pulsar.broker.topic.consumer.msg.ack
```
Comment (heesung-sn) [inline docs/reference-metrics-opentelemetry.md:53]: perhaps, pulsar.broker.topic.consumer.msg.ack.count

**Comment by dragosvictor** [comment]
> It would be great to clarify whether the metrics are accumulative or delta.

All these values are accumulative over time (say, absolute number of bytes). It is the duty of the backend stack to derive the deltas (such as bytes/second).

File: docs/reference-metrics-opentelemetry.md [hunk 24]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Unit: `{entry}`
+
+#### pulsar.broker.topic.compaction.compacted.entries.size
+The total size of the compacted entries.
+* Type: UpDownCounter
+* Unit: `{byte}`
+
+#### pulsar.broker.topic.transaction
+The number of transactions on this topic.
+* Type: UpDownCounter
```
[the remark this answers is not in the document]
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:150]: This metric has attributes for the tx status ([ref](https://github.com/apache/pulsar/pull/22467/files#diff-e3d8386fc3f9292780771f4eae19d9297637d374244f9c4b3f89b6c0e442e2e3R458-R470)): `active`, `committed`, `aborted`. The latter two can be counters, but not the first one.

File: docs/reference-metrics-opentelemetry.md [hunk 25]
```diff
@@ -8,6 +8,153 @@ Pulsar exposes the following OpenTelemetry metrics.
...
+* Unit: `{transaction}`
+
+#### pulsar.broker.topic.subscription.delayed
+The total message batches (entries) are delayed for dispatching.
+* Type: UpDownCounter
```
[the remark this answers is not in the document]
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:155]: This number can fluctuate up and down, it does not reflect a running total.

File: docs/reference-metrics-opentelemetry.md (hunk 3 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:11]: The implementation https://github.com/apache/pulsar/pull/22467 uses asynchronous instruments, due to performance and resource efficiency reasons (more details in that PR). This is a detail, however, so it doesn't need to be in the metrics reference documentation.

Furthermore, from the users point of view, this is transparent. Instrument types do not impact the way the user needs to collect or interpret these values.

File: docs/reference-metrics-opentelemetry.md (hunk 5 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:54]: This particular metric is not currently populated by the implementation. I'll remove it from the doc until it is.

File: docs/reference-metrics-opentelemetry.md (hunk 23 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:53]: Removed, as explained above.

File: docs/reference-metrics-opentelemetry.md (hunk 22 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:12]: There's some recommendations: https://opentelemetry.io/docs/specs/semconv/general/metrics/#use-count-instead-of-pluralization-for-updowncounters. I'll add the suffixes where suggested, sure!

File: docs/reference-metrics-opentelemetry.md (hunk 6 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md]: I don't see offloaded ledgers being subtracted from `ManagedLedgerImpl.totalSize` ([ref](https://github.com/apache/pulsar/blob/d5c72312ff4d03291e1ea2eb37464250c85bf401/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L190)). I suspect the value includes the offloaded numbers too, but I wasn't able to assess for certain. @merlimat @lhotari can you confirm?

File: docs/reference-metrics-opentelemetry.md (hunk 9 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md:87]: No, only the event counts, no information on the messages themselves. Perhaps we can add this as a follow-up 👍

File: docs/reference-metrics-opentelemetry.md (hunk 15 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md]: Refactored into just one metric named `pulsar.broker.topic.compaction.operation.count`, with an attribute to specify the result.

File: docs/reference-metrics-opentelemetry.md (hunk 14 above)
Comment (dragosvictor) [reply docs/reference-metrics-opentelemetry.md]: Refactored, as described [below](https://github.com/apache/pulsar-site/pull/880/files/8986d386d0f76d6a14e3cb709c801a053219a764#r1578260459).

**Review by heesung-sn — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**